File: RewardCore/Sources/RewardCore/Engine/ScoringEngine.swift
import Foundation

/// 计分相关的纯计算。所有方法只读输入、无副作用，便于确定性单测。
public enum ScoringEngine {

    /// 当前积分余额 = 有效事件分值之和 − 已通过兑换的花费之和。
    public static func balance(events: [ScoreEvent], redemptions: [RedemptionRequest]) -> Int {
        let earned = events.filter { !$0.isVoided }.reduce(0) { $0 + $1.points }
        let spent = redemptions.filter { $0.status == .approved }.reduce(0) { $0 + $1.cost }
        return earned - spent
    }

    /// 累计获得分（仅统计有效且为正的事件），用于里程碑徽章，不因花费/扣分而回退。
    public static func cumulativeEarned(events: [ScoreEvent]) -> Int {
        events.filter { !$0.isVoided && $0.points > 0 }.reduce(0) { $0 + $1.points }
    }

    /// 指定时间区间 [start, end) 内有效事件的净分。
    public static func netPoints(events: [ScoreEvent], in interval: DateInterval) -> Int {
        events
            .filter { !$0.isVoided && $0.timestamp >= interval.start && $0.timestamp < interval.end }
            .reduce(0) { $0 + $1.points }
    }

    /// 包含 now 的"本周"区间（依日历 firstWeekday 决定周起点）。
    public static func currentWeekInterval(now: Date, calendar: Calendar = .current) -> DateInterval {
        calendar.dateInterval(of: .weekOfYear, for: now)
            ?? DateInterval(start: now, duration: 7 * 86_400)
    }

    /// 最近一条可撤销（未作废）事件——按时间倒序取第一条。
    public static func lastUndoableEvent(events: [ScoreEvent]) -> ScoreEvent? {
        events.filter { !$0.isVoided }.max { $0.timestamp < $1.timestamp }
    }

    /// 最近的有效事件，按时间倒序，限制条数。
    public static func recentEvents(events: [ScoreEvent], limit: Int) -> [ScoreEvent] {
        Array(events.filter { !$0.isVoided }
            .sorted { $0.timestamp > $1.timestamp }
            .prefix(max(0, limit)))
    }

    /// 最近 `days` 天每天的净分，最旧在前。每个 date 为当天 00:00。
    public static func dailyNetTotals(
        events: [ScoreEvent],
        days: Int,
        endingOn: Date,
        calendar: Calendar = .current
    ) -> [DailyScore] {
        guard days > 0 else { return [] }
        let endDay = calendar.startOfDay(for: endingOn)
        let active = events.filter { !$0.isVoided }
        return (0..<days).reversed().compactMap { back -> DailyScore? in
            guard let dayStart = calendar.date(byAdding: .day, value: -back, to: endDay) else { return nil }
            let net = active
                .filter { calendar.isDate($0.timestamp, inSameDayAs: dayStart) }
                .reduce(0) { $0 + $1.points }
            return DailyScore(date: dayStart, net: net)
        }
    }
}

File: web/src/domain/scoringEngine.ts
// ScoringEngine 移植：纯函数、确定性；voided 事件在一切计算中排除。

import type { DailyScore, DateInterval, RedemptionRequest, ScoreEvent } from './types'
import { addDays, isSameDay, startOfDay, weekInterval } from './dates'

/** 余额 = Σ(非 voided 事件分) − Σ(approved 兑换花费)。可为负。 */
export function balance(events: ScoreEvent[], redemptions: RedemptionRequest[]): number {
  const earned = events.reduce((sum, e) => (e.isVoided ? sum : sum + e.points), 0)
  const spent = redemptions.reduce((sum, r) => (r.status === 'approved' ? sum + r.cost : sum), 0)
  return earned - spent
}

/** 累计获得：只算非 voided 且 points > 0 的事件（里程碑徽章用，永不因扣分/兑换回退）。 */
export function cumulativeEarned(events: ScoreEvent[]): number {
  return events.reduce((sum, e) => (!e.isVoided && e.points > 0 ? sum + e.points : sum), 0)
}

/** 区间净分：[start, end) 含头不含尾。 */
export function netPoints(events: ScoreEvent[], interval: DateInterval): number {
  return events.reduce((sum, e) => {
    if (e.isVoided) return sum
    if (e.timestamp >= interval.start && e.timestamp < interval.end) return sum + e.points
    return sum
  }, 0)
}

/** 当前周区间（默认周一起始）。 */
export function currentWeekInterval(now: Date, firstWeekday: number = 1): DateInterval {
  return weekInterval(now, firstWeekday)
}

/** 最近一条可撤销（非 voided）事件；时间戳最大者，同刻取后者（对位 Swift max(by:)）。 */
export function lastUndoableEvent(events: ScoreEvent[]): ScoreEvent | null {
  let latest: ScoreEvent | null = null
  for (const e of events) {
    if (e.isVoided) continue
    if (latest === null || e.timestamp.getTime() >= latest.timestamp.getTime()) latest = e
  }
  return latest
}

/** 最近事件：非 voided，按时间降序，取前 max(0, limit) 条。 */
export function recentEvents(events: ScoreEvent[], limit: number): ScoreEvent[] {
  return events
    .filter((e) => !e.isVoided)
    .sort((a, b) => b.timestamp.getTime() - a.timestamp.getTime())
    .slice(0, Math.max(0, limit))
}

/** 每日净分：最近 days 天（含 endingOn 当天），最旧在前；无事件的天 net=0 也占位。 */
export function dailyNetTotals(events: ScoreEvent[], days: number, endingOn: Date): DailyScore[] {
  if (days <= 0) return []
  const endDay = startOfDay(endingOn)
  const active = events.filter((e) => !e.isVoided)
  const result: DailyScore[] = []
  for (let offset = 0; offset < days; offset++) {
    const daysBack = days - 1 - offset
    // startOfDay 重归一化：跨「本地午夜不存在」的 DST 日时 addDays 会偏移到 01:00
    const dayStart = startOfDay(addDays(endDay, -daysBack))
    const net = active.reduce((sum, e) => (isSameDay(e.timestamp, dayStart) ? sum + e.points : sum), 0)
    result.push({ date: dayStart, net })
  }
  return result
}
